### DISCUSSION TITLE
Week 4: ML with Python

### DISCUSSION THREAD

**Discussion by BekahHW** [deepgram/community#89] [General help]
## Getting Started

The goal this week is to get to "Create a Play Generator" lesson from [freeCodeCamp's ML with python](https://www.freecodecamp.org/learn/machine-learning-with-python/) and check-in here. 

According to the [Deepgram Blog](https://blog.deepgram.com/how-to-speak-ai-a-beginners-dictionary/#natural-language-processing-(nlp)), NLP is:

> The branch of AI dedicated to getting computers to understand language (and all of the complexities, rule-exceptions, and patterns these languages entail). This is the intersection of artificial intelligence and linguistics.

> NLP is the reason why your Siri and Alexa can understand you when you talk, why Google Translate is becoming increasingly useful, why search engines understand your questions, why AI can now write code for you, and much more.

## Learning Resources
- [How to Speak AI: A Beginner's Dictionary](https://blog.deepgram.com/how-to-speak-ai-a-beginners-dictionary/)

### Sentiment Analysis:
- https://blog.deepgram.com/sentiment-analysis-emotion-regulation-difference/

### Recurrent Neural Network
- [What are Recurrent Neural Networks](https://www.ibm.com/topics/recurrent-neural-networks)
- [An Introduction to Recurrent Neural Networks and the Math that Powers Them](https://machinelearningmastery.com/an-introduction-to-recurrent-neural-networks-and-the-math-that-powers-them/)

## Check-in
1. What have you learned about Recurrent neural networks? 
2. What are some use cases for RNNs?
3. What do you know today that you didn't when you started?
4. What are some key terms or ideas from these lessons?
5. What do you need to be clarified or what questions do you have?
6. What are you excited to learn more about?

**Comment by BekahHW** [comment]
This resource seems to go along with the video: https://medium.com/geekculture/natural-language-processing-ed48334f80be

1. What have you learned about Recurrent neural networks? A type of neural network that shares the same weight parameter across each layer of the network. Takes order into account.
2. What are some use cases for RNNs? NLP, speech recognition, captioning, translation, and sentiment analysis as a subset.
3. What do you know today that you didn't when you started? I wasn't familiar with the term word embeddings.
4. What are some key terms or ideas from these lessons? word embeddings, long short term memory.
5. What do you need to be clarified or what questions do you have?
6. What are you excited to learn more about? This feels a lot more comfortable to me since NLP features are things that we talk a lot about at DG. It would be fun to dive more into feature understanding like sentiment, paragraphs, and summarization.

**Comment by paceaux** [comment]
FYI

Anyone wanting to run the [RNN Colab](https://colab.research.google.com/drive/1ysEKrw_LE2jMndo1snrZUh5w87LQsCxk#forceEdit=true&sandboxMode=true)  should be advised that it won't work without a small change to code blocks 12 and 13. 

block 12:

```python
from tensorflow.keras.preprocessing.sequence import pad_sequences
``` 

block 13

```python
train_data = pad_sequences(train_data, MAXLEN)
test_data = pad_sequences(test_data, MAXLEN)
```

**Comment by paceaux** [comment]
1. What did I learn about RNNs? They're "recurrent". A loop. They don't process the whole thing at once. Instead it processes at steps and each step changes the data. So it maintains an internal memory and each input changes the context of the previous

2. Use-casses for RNN? Sentiment Analysis. Text processing. _Language identification_. Profiling. 
3. What do I know today that I didn't know when I started. **IT'S STILL NUMBERS**. That seems a bit silly, to be honest. But apparently this still requires encoding your information into integers and then dealing with the integers. That just feels... _wrong_
4. Key terms and ideas?
    * Word embeddings
    * preprocessing
    * LSTM
5. What needs to be clarified? I'm a bit fuzzy on this vectorification of words. Also... **WHY NUMBERS AGAIN**? Also... Why does the play generator example require converting each individual character into a number
6. Excited to learn about? How to actually predict stuff with text. I almost can see how to do bigram/ngram analysis with this.
